### ISSUE TITLE
Per-FedAVG clients are trained twice

### ISSUE THREAD

**Issue by likeyhnbm** [TsingZ0/PFL-Non-IID#59]
I found there will be two `client.train()` called in each round. Is it on purpose? Could you please explain it to me? Thanks!

**Comment by NJU-TJL** [comment]
I have the same question.
In `serverperavg.py`:

```
            # choose several clients to send back upated model to server
            for client in self.selected_clients:
                client.train()
                client.train()
```

**Comment by NJU-TJL** [comment]
> I found there will be two `client.train()` called in each round. Is it on purpose? Could you please explain it to me? Thanks!

Perhaps because of this, you can see the commit record this time:
https://github.com/TsingZ0/PFL-Non-IID/commit/4f394efe04f30dbd3cab4278467631854f997903#diff-63e160f53ad3e0f07661ac69c3db7896c4d81fb8396e705d38ac1d680e175e41

Note the difference in the train() function, batch_size*2, so we need to train() twice.

**Comment by TsingZ0** [comment]
@NJU-TJL Yes. 

I call `client.train()` twice to keep the number of update steps of Per-FedAvg the same as other methods, e.g., FedAvg. To perform meta-learning for Per-FedAvg, I first update the local model to an intermediate model with the first batch of data, obtain the gradients of the intermediate model on the second batch of data, and update the original local model with these gradients. Thus, only half of the data in the training set updates the local model if one calls `client.train()` once.

**Comment by Flying7514** [comment]
How is the Hessian matrix of Per-FedAvg calculated?
